In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
cd fahrgastzahlen_2019

[Errno 2] No such file or directory: 'fahrgastzahlen_2019'
/Users/home/Ironhack/Smart_Cities_for_Public_Transport/notebooks


In [3]:
lines = pd.read_csv('../raw_data/LINIE.csv', sep=',')
lines = lines.Linienname.unique()

In [4]:
lines

array([ 10,  11,  12,  13,  14,  15, 161, 162, 163, 165,  17, 184, 185,
         2,  25,  29,   3, 301, 302, 303, 304, 305, 306, 307, 308, 309,
        31, 311, 314, 315,  32, 325,  33, 330, 331, 341, 342, 343, 344,
       345, 346, 347, 348, 349,  35, 352,  37,  38,  39, 391, 392, 394,
       395, 396, 397, 398,   4,  40, 411, 412, 417, 418, 419,  45,  46,
         5,  54,   6,  61,  62,  64,  66,  67,  69,   7,  70, 701, 703,
       704, 705,  72,  73, 743, 744, 745,  75, 751, 753,  76,  77,  78,
        79,   8,  80,  83,  89,   9,  91, 910, 912, 916, 917, 918, 919,
        94])

In [5]:
zh = pd.read_csv('../clean_data/zurich_transport.csv')

In [6]:
zh

Unnamed: 0  daytype  line_name  ride_id  stop_id  next_stop_id  \
0                 0        0          2    73477      190         165.0   
1                 1        0          2    73477      165         166.0   
2                 2        0          2    73477      166         167.0   
3                 3        0          2    73477      167         168.0   
4                 4        0          2    73477      168         216.0   
...             ...      ...        ...      ...      ...           ...   
1020709     1249459        0          2   119565      695         696.0   
1020710     1249460        0          2   119565      696         697.0   
1020711     1249461        0          2   119565      697         293.0   
1020712     1249462        0          2   119565      293         771.0   
1020713     1249463        0          2   119565      771           NaN   

             time  number_of_measurements  occupancy  days_measured  \
0        17:45:30                       5      83.20          34.15   
1        17:47:30                       5      83.40          34.15   
2        17:48:36                       5      86.40          34.15   
3        17:50:30                       5      87.80          34.15   
4        17:52:30                       5      67.80          34.15   
...           ...                     ...        ...            ...   
1020709  05:36:42                       4       7.50          13.84   
1020710  05:38:06                       4       7.75          13.84   
1020711  05:39:12                       4       9.00          13.84   
1020712  05:40:30                       4       1.75          13.84   
1020713  05:41:06                       4        NaN          13.84   

        vehicle_type  maximum_capacity  occupancy_rate  
0               tram               220        0.378182  
1               tram               220        0.379091  
2               tram               220        0.392727  
3               tram               220        0.399091  
4               tram               220        0.308182  
...              ...               ...             ...  
1020709         tram               220        0.034091  
1020710         tram               220        0.035227  
1020711         tram               220        0.040909  
1020712         tram               220        0.007955  
1020713         tram               220             NaN  

[1020714 rows x 13 columns]

In [7]:
stops = pd.read_csv('../raw_data/HALTESTELLEN.csv')

In [8]:
stops

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
0                724                2277                 SHDO   
1                377                3822                 BKIS   
2                767                1397                 KIND   
3                766                 757                 EIHH   
4                347                6506               BABH07   
..               ...                 ...                  ...   
759              120                3011                 ZWEI   
760              138                3015                 ZWIE   
761              234                3018                 ZWIN   
762              229                3019                 ZYPR   
763              227                1357                 KALK   

               Haltestellenlangname\t  
0     Kilchberg ZH\tSchulhaus Dorfstr  
1                Killwangen\t Bahnhof  
2                     Kindhausen AG\t  
3            Kindhausen AG\t Eichholz  
4           Kloten Balsberg\t Bahnhof  
..                                ...  
759        Zürich\t Zweiackerstrasse  
760              Zürich\t Zwielplatz  
761             Zürich\t Zwinglihaus  
762        Zürich\t Zypressenstrasse  
763  Zürich\tKalkbreite/Bhf.Wiedikon  

[764 rows x 4 columns]

# First analysis of place & time where vehicles are the fullest

In [9]:
# I will now do some first queries. I want to know which ride has the highest average amount of people for trams and buses
# What time is this at? Where in the city? What is the difference between weekdays and weekends?
# Where do we have vehicles that are constantly being underused and could be better used elswhere?

zh.sort_values(by=['occupancy_rate'],ascending=False).head(15)

Unnamed: 0  daytype  line_name  ride_id  stop_id  next_stop_id  \
357750      362748        0        330     2957      283         282.0   
357751      362749        0        330     2957      282         280.0   
982250     1210884        0        302   105802      293         294.0   
326573      331562        0         89     7012       14          13.0   
326572      331561        0         89     7012       15          14.0   
326571      331560        0         89     7012       16          15.0   
344015      349008        0        303    32549      358         274.0   
153188      158080        1         62    90883      156         496.0   
368374      373372        0         83    59134       60          59.0   
368326      373324        0         83    42148       60          59.0   
368310      373308        0         83    37043       60          59.0   
368375      373373        0         83    59134       59          81.0   
368329      373327        0         83    42148       82          25.0   
161833      166768        0        418    21311      169         108.0   
161834      166769        0        418    21311      108         110.0   

            time  number_of_measurements  occupancy  days_measured  \
357750  11:12:00                       1      97.00         188.00   
357751  11:13:30                       1      97.00         188.00   
982250  13:29:06                       1      96.00          73.82   
326573  08:16:36                      49      75.81         251.00   
326572  08:15:36                      49      75.61         251.00   
326571  08:14:30                      49      75.18         251.00   
344015  07:05:06                      13      73.84         177.17   
153188  17:26:12                       3      73.00          31.39   
368374  08:30:54                      55      72.85         251.00   
368326  08:08:24                      47      72.82         251.00   
368310  08:00:54                      40      72.80         251.00   
368375  08:32:06                      55      72.12         251.00   
368329  08:12:30                      47      71.72         251.00   
161833  07:33:00                      59     196.57         203.00   
161834  07:34:48                      59     196.47         203.00   

       vehicle_type  maximum_capacity  occupancy_rate  
357750    large_bus               100        0.970000  
357751    large_bus               100        0.970000  
982250    large_bus               100        0.960000  
326573          bus                80        0.947625  
326572          bus                80        0.945125  
326571          bus                80        0.939750  
344015          bus                80        0.923000  
153188          bus                80        0.912500  
368374          bus                80        0.910625  
368326          bus                80        0.910250  
368310          bus                80        0.910000  
368375          bus                80        0.901500  
368329          bus                80        0.896500  
161833         tram               220        0.893500  
161834         tram               220        0.893045

In [10]:
zh.loc[zh.occupancy_rate > 0.85].sort_values(by=['occupancy_rate'],ascending=False)

Unnamed: 0  daytype  line_name  ride_id  stop_id  next_stop_id  \
357750       362748        0        330     2957      283         282.0   
357751       362749        0        330     2957      282         280.0   
982250      1210884        0        302   105802      293         294.0   
326573       331562        0         89     7012       14          13.0   
326572       331561        0         89     7012       15          14.0   
326571       331560        0         89     7012       16          15.0   
344015       349008        0        303    32549      358         274.0   
153188       158080        1         62    90883      156         496.0   
368374       373372        0         83    59134       60          59.0   
368326       373324        0         83    42148       60          59.0   
368310       373308        0         83    37043       60          59.0   
368375       373373        0         83    59134       59          81.0   
368329       373327        0         83    42148       82          25.0   
161833       166768        0        418    21311      169         108.0   
161834       166769        0        418    21311      108         110.0   
368327       373325        0         83    42148       59          81.0   
153187       158079        1         62    90883      155         156.0   
1013851     1242576        0          2   100054       81         305.0   
271626       276585        0         62    78716      155         156.0   
161622       166557        0        418    21310      110         108.0   
271627       276586        0         62    78716      156         496.0   
402323       407374        0        418    35657      110         108.0   
161621       166556        0        418    21310      173         110.0   
345985       350978        0        910     2284      526         527.0   
326625       331614        0         89    64441       16          15.0   
326627       331616        0         89    64441       14          13.0   
423786       428846        0         80    73021      133         134.0   
368294       373292        0         83    31462       60          59.0   
326626       331615        0         89    64441       15          14.0   
402322       407373        0        418    35657      173         110.0   
161635       166570        0        418     3040      110         108.0   
980283      1208917        0        303   104193      357         274.0   
701030       706559        1         62    29671      155         156.0   
701031       706560        1         62    29671      156         496.0   
963904      1191934        0          6   114424      256         192.0   

             time  number_of_measurements  occupancy  days_measured  \
357750   11:12:00                       1      97.00         188.00   
357751   11:13:30                       1      97.00         188.00   
982250   13:29:06                       1      96.00          73.82   
326573   08:16:36                      49      75.81         251.00   
326572   08:15:36                      49      75.61         251.00   
326571   08:14:30                      49      75.18         251.00   
344015   07:05:06                      13      73.84         177.17   
153188   17:26:12                       3      73.00          31.39   
368374   08:30:54                      55      72.85         251.00   
368326   08:08:24                      47      72.82         251.00   
368310   08:00:54                      40      72.80         251.00   
368375   08:32:06                      55      72.12         251.00   
368329   08:12:30                      47      71.72         251.00   
161833   07:33:00                      59     196.57         203.00   
161834   07:34:48                      59     196.47         203.00   
368327   08:09:36                      47      71.31         251.00   
153187   17:24:54                       3      71.00          31.39   
1013851  17:40:12             

### There seem to be two timeperiods in which the occupancy in the public transport system is the highest: Between 07:00 - 08:00 and 17:00 - 18:00. 
### The clear assumption here is that these are the rush hour periods of people going to and from work, especially since they are all during weekdays. 

# Overused Stops Morning

In [11]:
# now I will determine all the vehicles that have an occupancy rate higher than 85% in the morning

zh.loc[(zh.time > '07:00:00') & (zh.time < '09:00:00') & (zh.occupancy_rate > 0.85)].sort_values(by=['occupancy_rate'],ascending=False)

Unnamed: 0  daytype  line_name  ride_id  stop_id  next_stop_id  \
326573      331562        0         89     7012       14          13.0   
326572      331561        0         89     7012       15          14.0   
326571      331560        0         89     7012       16          15.0   
344015      349008        0        303    32549      358         274.0   
368374      373372        0         83    59134       60          59.0   
368326      373324        0         83    42148       60          59.0   
368310      373308        0         83    37043       60          59.0   
368375      373373        0         83    59134       59          81.0   
368329      373327        0         83    42148       82          25.0   
161833      166768        0        418    21311      169         108.0   
161834      166769        0        418    21311      108         110.0   
368327      373325        0         83    42148       59          81.0   
161622      166557        0        418    21310      110         108.0   
402323      407374        0        418    35657      110         108.0   
161621      166556        0        418    21310      173         110.0   
326625      331614        0         89    64441       16          15.0   
326627      331616        0         89    64441       14          13.0   
423786      428846        0         80    73021      133         134.0   
368294      373292        0         83    31462       60          59.0   
326626      331615        0         89    64441       15          14.0   
402322      407373        0        418    35657      173         110.0   
161635      166570        0        418     3040      110         108.0   
980283     1208917        0        303   104193      357         274.0   
963904     1191934        0          6   114424      256         192.0   

            time  number_of_measurements  occupancy  days_measured  \
326573  08:16:36                      49      75.81         251.00   
326572  08:15:36                      49      75.61         251.00   
326571  08:14:30                      49      75.18         251.00   
344015  07:05:06                      13      73.84         177.17   
368374  08:30:54                      55      72.85         251.00   
368326  08:08:24                      47      72.82         251.00   
368310  08:00:54                      40      72.80         251.00   
368375  08:32:06                      55      72.12         251.00   
368329  08:12:30                      47      71.72         251.00   
161833  07:33:00                      59     196.57         203.00   
161834  07:34:48                      59     196.47         203.00   
368327  08:09:36                      47      71.31         251.00   
161622  07:17:24                      59     193.74         203.00   
402323  07:17:24                      14     193.14          48.00   
161621  07:14:18                      59     192.35         203.00   
326625  08:45:00                      40      69.85         251.00   
326627  08:47:06                      40      69.80         251.00   
423786  08:32:18                       6      87.00          48.00   
368294  07:53:24                      52      69.44         251.00   
326626  08:46:06                      40      69.20         251.00   
402322  07:14:18                      14     189.57          48.00   
161635  07:32:24                      29     188.65         203.00   
980283  07:18:36                       5      68.40          73.82   
963904  07:46:54                       9     187.44          59.67   

       vehicle_type  maximum_capacity  occupancy_rate  
326573          bus                80        0.947625  
326572          bus                80        0.945125  
326571          bus                80        0.939750  
344015          bus                80        0.923000  
368374          bus                80        0.910625  
368326          bus                80        0.910250  
368310        

In [12]:
overused_morning = zh[zh.occupancy_rate > 0.85].loc[(zh.time > '07:00:00') & (zh.time < '09:00:00')].sort_values(
    by='occupancy_rate', ascending=False)

overused_morning.line_name.value_counts()

83     7
418    7
89     6
303    2
80     1
6      1
Name: line_name, dtype: int64

In [13]:
overused_morning.stop_id.value_counts()

60     4
110    3
59     2
16     2
15     2
14     2
173    2
82     1
108    1
357    1
169    1
358    1
133    1
256    1
Name: stop_id, dtype: int64

## Overused Evening

In [14]:
overused_evening = zh[zh.occupancy_rate > 0.85].loc[(zh.time > '17:00:00') & (zh.time < '18:00:00')].sort_values(by=['occupancy_rate']
                                                                    ,ascending=False)
overused_evening.line_name.value_counts()

62    4
2     1
Name: line_name, dtype: int64

In [15]:
overused_evening

Unnamed: 0  daytype  line_name  ride_id  stop_id  next_stop_id  \
153188       158080        1         62    90883      156         496.0   
153187       158079        1         62    90883      155         156.0   
1013851     1242576        0          2   100054       81         305.0   
271626       276585        0         62    78716      155         156.0   
271627       276586        0         62    78716      156         496.0   

             time  number_of_measurements  occupancy  days_measured  \
153188   17:26:12                       3      73.00          31.39   
153187   17:24:54                       3      71.00          31.39   
1013851  17:40:12                       1     195.00          13.84   
271626   17:32:54                      34      70.85         203.00   
271627   17:34:24                      34      70.41         203.00   

        vehicle_type  maximum_capacity  occupancy_rate  
153188           bus                80        0.912500  
153187           bus                80        0.887500  
1013851         tram               220        0.886364  
271626           bus                80        0.885625  
271627           bus                80        0.880125

In [16]:
overused_evening.stop_id.value_counts()

156    2
155    2
81     1
Name: stop_id, dtype: int64

# Overused Stops Morning

In [17]:
# I will now use the Stop ID from the main table to search for the name of the overused stops 
# in order to be able to locate them geographically

In [18]:
stops[stops.Haltestellen_Id == 60]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
210               60                3029                 HSBB   

            Haltestellenlangname\t  
210  Zürich\t Bahnhof Hardbrücke

In [19]:
stops[stops.Haltestellen_Id == 110]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
344              110                1144                 HEGI   

      Haltestellenlangname\t  
344  Zürich\t Hegibachplatz

In [20]:
stops[stops.Haltestellen_Id == 59]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
339               59                1122                 HARP   

    Haltestellenlangname\t  
339    Zürich\t Hardplatz

In [21]:
stops[stops.Haltestellen_Id == 16]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
206               16                3022                 BALT   

           Haltestellenlangname\t  
206  Zürich\t Bahnhof Altstetten

In [22]:
stops[stops.Haltestellen_Id == 15]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
252               15                 542                 BRIS   

       Haltestellenlangname\t  
252  Zürich\t Bristenstrasse

In [23]:
stops[stops.Haltestellen_Id == 14]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
227               14                 427                 BASL   

      Haltestellenlangname\t  
227  Zürich\t Baslerstrasse

In [24]:
stops[stops.Haltestellen_Id == 173]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
225              173                 421                 BALG   

    Haltestellenlangname\t  
225     Zürich\t Balgrist

In [25]:
stops[stops.Haltestellen_Id == 82]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
191               82                1446                 AKRE   

          Haltestellenlangname\t  
191  Zürich\t Altes Krematorium

In [26]:
stops[stops.Haltestellen_Id == 108]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
399              108                1453                 KREU   

    Haltestellenlangname\t  
399   Zürich\t Kreuzplatz

In [27]:
stops[stops.Haltestellen_Id == 357]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
453              357                1843                 OETW   

         Haltestellenlangname\t  
453  Dietikon\t Oetwilerstrasse

In [28]:
stops[stops.Haltestellen_Id == 169]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
216              169                3039                 BSTA   

            Haltestellenlangname\t  
216  Zürich\t Bahnhof Stadelhofen

In [29]:
stops[stops.Haltestellen_Id == 358]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
464              358                3798                 VORS   

    Haltestellenlangname\t  
464    Dietikon\t Vorstadt

In [30]:
stops[stops.Haltestellen_Id == 133]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
564              133                1539                 LERC   

     Haltestellenlangname\t  
564  Zürich\t Lerchenhalde

In [31]:
stops[stops.Haltestellen_Id == 256]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
336              256                1100                 HEGG   

    Haltestellenlangname\t  
336    Zürich\t Haldenegg

# Overused Stops Evening

In [32]:
stops[stops.Haltestellen_Id == 156]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
626              156                2045                 REGB   

          Haltestellenlangname\t  
626  Zürich\t Regensbergbrücke

In [33]:
stops[stops.Haltestellen_Id == 155]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
212              155                3034                 BOER   

         Haltestellenlangname\t  
212  Zürich\t Bahnhof Oerlikon

In [34]:
stops[stops.Haltestellen_Id == 305]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
568              305                1548                 LETG   

    Haltestellenlangname\t  
568   Zürich\t Letzigrund

In [35]:
stops[stops.Haltestellen_Id == 102]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
680              102                1366                 SIHP   

      Haltestellenlangname\t  
680  Zürich\t Sihlpost / HB

In [36]:
stops[stops.Haltestellen_Id == 81]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
188               81                  48                 ALBP   

         Haltestellenlangname\t  
188  Zürich\t Albisriederplatz

In [37]:
stops[stops.Haltestellen_Id == 358]

Haltestellen_Id  Haltestellennummer Haltestellenkurzname  \
464              358                3798                 VORS   

    Haltestellenlangname\t  
464    Dietikon\t Vorstadt

# Vehicles with low occupancy

### I will now try to find the vehicles with the lowest maximum occupancy (lower than 50%) during the two timeframes that I am analyzing. That way I will be able to to divert some of these vehicles and towards the overloaded lines, so as to better make use of the capacities of the Zurich public transport network. 

### Create a dictionary to find all the lines that never go above 50 % during the overused hours

In [38]:
# lets start with buses
mydict_bus = {}
for line in lines:
    mydict_bus[line] = zh[(zh.line_name == line) & (zh.daytype == 0) & (zh.vehicle_type == 'bus')].loc[(zh.time > '07:00:00') & 
                        (zh.time < '09:00:00')].sort_values(by=['occupancy_rate']).occupancy_rate.max()
    

In [39]:
mydict_bus

{10: nan,
 11: nan,
 12: nan,
 13: nan,
 14: nan,
 15: nan,
 161: 0.674625,
 162: 0.3615,
 163: 0.102125,
 165: 0.63775,
 17: nan,
 184: 0.48675,
 185: 0.481125,
 2: nan,
 25: 0.30475,
 29: nan,
 3: nan,
 301: 0.7255,
 302: nan,
 303: 0.9229999999999999,
 304: nan,
 305: 0.26375000000000004,
 306: 0.7228749999999999,
 307: 0.149625,
 308: nan,
 309: 0.725625,
 31: nan,
 311: 0.4365,
 314: 0.363,
 315: 0.46425,
 32: nan,
 325: 0.4625,
 33: nan,
 330: nan,
 331: nan,
 341: nan,
 342: nan,
 343: nan,
 344: nan,
 345: nan,
 346: nan,
 347: nan,
 348: nan,
 349: nan,
 35: 0.390625,
 352: nan,
 37: 0.650875,
 38: 0.1075,
 39: 0.21,
 391: nan,
 392: nan,
 394: nan,
 395: nan,
 396: nan,
 397: nan,
 398: nan,
 4: nan,
 40: 0.236875,
 411: nan,
 412: nan,
 417: nan,
 418: nan,
 419: nan,
 45: nan,
 46: nan,
 5: nan,
 54: 0.41475,
 6: nan,
 61: 0.81925,
 62: 0.812125,
 64: 0.375125,
 66: 0.5615,
 67: 0.622125,
 69: 0.68425,
 7: nan,
 70: 0.63525,
 701: 0.78175,
 703: 0.64325,
 704: 0.65775,
 705

In [40]:
# trams are next
mydict_tram = {}
for line in lines:
    mydict_tram[line] = zh[(zh.line_name == line) & (zh.daytype == 0) & (zh.vehicle_type == 'tram')].loc[(zh.time > '07:00:00') & 
                        (zh.time < '09:00:00')].sort_values(by=['occupancy_rate']).occupancy_rate.max()

In [41]:
mydict_tram

{10: 0.6504545454545454,
 11: 0.6900454545454545,
 12: 0.6652272727272727,
 13: 0.6950000000000001,
 14: 0.5333181818181818,
 15: 0.2912272727272727,
 161: nan,
 162: nan,
 163: nan,
 165: nan,
 17: 0.7623181818181819,
 184: nan,
 185: nan,
 2: 0.7196818181818182,
 25: nan,
 29: nan,
 3: 0.6028181818181818,
 301: nan,
 302: nan,
 303: nan,
 304: nan,
 305: nan,
 306: nan,
 307: nan,
 308: nan,
 309: nan,
 31: nan,
 311: nan,
 314: nan,
 315: nan,
 32: nan,
 325: nan,
 33: nan,
 330: nan,
 331: nan,
 341: nan,
 342: nan,
 343: nan,
 344: nan,
 345: nan,
 346: nan,
 347: nan,
 348: nan,
 349: nan,
 35: nan,
 352: nan,
 37: nan,
 38: nan,
 39: nan,
 391: nan,
 392: nan,
 394: nan,
 395: nan,
 396: nan,
 397: nan,
 398: nan,
 4: 0.7242727272727273,
 40: nan,
 411: nan,
 412: nan,
 417: 0.7850909090909091,
 418: 0.8935,
 419: nan,
 45: nan,
 46: nan,
 5: 0.7071818181818182,
 54: nan,
 6: 0.852,
 61: nan,
 62: nan,
 64: nan,
 66: nan,
 67: nan,
 69: nan,
 7: 0.7568181818181818,
 70: nan,
 70

In [42]:
mydict_large_bus = {}
for line in lines:
    mydict_large_bus[line] = zh[(zh.line_name == line) & (zh.daytype == 0) & (zh.vehicle_type == 'large_bus')].loc[(zh.time > '07:00:00') & 
                        (zh.time < '09:00:00')].sort_values(by=['occupancy_rate']).occupancy_rate.max()

In [43]:
mydict_large_bus

{10: nan,
 11: nan,
 12: nan,
 13: nan,
 14: nan,
 15: nan,
 161: nan,
 162: nan,
 163: nan,
 165: nan,
 17: nan,
 184: nan,
 185: nan,
 2: nan,
 25: nan,
 29: nan,
 3: nan,
 301: nan,
 302: 0.7066,
 303: nan,
 304: 0.7837999999999999,
 305: nan,
 306: nan,
 307: nan,
 308: 0.8195,
 309: nan,
 31: nan,
 311: nan,
 314: nan,
 315: nan,
 32: nan,
 325: nan,
 33: nan,
 330: 0.3531,
 331: nan,
 341: nan,
 342: nan,
 343: nan,
 344: nan,
 345: nan,
 346: nan,
 347: nan,
 348: nan,
 349: nan,
 35: nan,
 352: nan,
 37: nan,
 38: nan,
 39: nan,
 391: nan,
 392: nan,
 394: nan,
 395: nan,
 396: nan,
 397: nan,
 398: nan,
 4: nan,
 40: nan,
 411: nan,
 412: nan,
 417: nan,
 418: nan,
 419: nan,
 45: nan,
 46: nan,
 5: nan,
 54: nan,
 6: nan,
 61: nan,
 62: nan,
 64: nan,
 66: nan,
 67: nan,
 69: nan,
 7: nan,
 70: nan,
 701: nan,
 703: nan,
 704: nan,
 705: nan,
 72: nan,
 73: nan,
 743: nan,
 744: nan,
 745: nan,
 75: nan,
 751: nan,
 753: nan,
 76: nan,
 77: nan,
 78: nan,
 79: nan,
 8: nan,
 

In [44]:
mydict_trolleybus = {}
for line in lines:
    mydict_trolleybus[line] = zh[(zh.line_name == line) & (zh.daytype == 0) & (zh.vehicle_type == 'trolleybus')].loc[(zh.time > '07:00:00') & 
                        (zh.time < '09:00:00')].sort_values(by=['occupancy_rate']).occupancy_rate.max()

In [45]:
mydict_trolleybus

{10: nan,
 11: nan,
 12: nan,
 13: nan,
 14: nan,
 15: nan,
 161: nan,
 162: nan,
 163: nan,
 165: nan,
 17: nan,
 184: nan,
 185: nan,
 2: nan,
 25: nan,
 29: nan,
 3: nan,
 301: nan,
 302: nan,
 303: nan,
 304: nan,
 305: nan,
 306: nan,
 307: nan,
 308: nan,
 309: nan,
 31: 0.74,
 311: nan,
 314: nan,
 315: nan,
 32: 0.8153846153846154,
 325: nan,
 33: 0.6,
 330: nan,
 331: nan,
 341: nan,
 342: nan,
 343: nan,
 344: nan,
 345: nan,
 346: nan,
 347: nan,
 348: nan,
 349: nan,
 35: nan,
 352: nan,
 37: nan,
 38: nan,
 39: nan,
 391: nan,
 392: nan,
 394: nan,
 395: nan,
 396: nan,
 397: nan,
 398: nan,
 4: nan,
 40: nan,
 411: 0.4524615384615385,
 412: 0.2746923076923077,
 417: nan,
 418: nan,
 419: nan,
 45: nan,
 46: 0.5756153846153846,
 5: nan,
 54: nan,
 6: nan,
 61: nan,
 62: nan,
 64: nan,
 66: nan,
 67: nan,
 69: nan,
 7: nan,
 70: nan,
 701: nan,
 703: nan,
 704: nan,
 705: nan,
 72: 0.6181538461538462,
 73: nan,
 743: nan,
 744: nan,
 745: nan,
 75: nan,
 751: nan,
 753: nan

## Do the same analysis for underused vehicles in the evening

In [46]:
mydict_large_bus = {}
for line in lines:
    mydict_large_bus[line] = zh[(zh.line_name == line) & (zh.daytype == 0) & (zh.vehicle_type == 'large_bus')].loc[(zh.time > '17:00:00') & 
                        (zh.time < '18:00:00')].sort_values(by=['occupancy_rate']).occupancy_rate.max()

In [47]:
mydict_large_bus

{10: nan,
 11: nan,
 12: nan,
 13: nan,
 14: nan,
 15: nan,
 161: nan,
 162: nan,
 163: nan,
 165: nan,
 17: nan,
 184: nan,
 185: nan,
 2: nan,
 25: nan,
 29: nan,
 3: nan,
 301: nan,
 302: 0.62,
 303: nan,
 304: 0.6406000000000001,
 305: nan,
 306: nan,
 307: nan,
 308: 0.6046,
 309: nan,
 31: nan,
 311: nan,
 314: nan,
 315: nan,
 32: nan,
 325: nan,
 33: nan,
 330: nan,
 331: nan,
 341: nan,
 342: nan,
 343: nan,
 344: nan,
 345: nan,
 346: nan,
 347: nan,
 348: nan,
 349: nan,
 35: nan,
 352: nan,
 37: nan,
 38: nan,
 39: nan,
 391: nan,
 392: nan,
 394: nan,
 395: nan,
 396: nan,
 397: nan,
 398: nan,
 4: nan,
 40: nan,
 411: nan,
 412: nan,
 417: nan,
 418: nan,
 419: nan,
 45: nan,
 46: nan,
 5: nan,
 54: nan,
 6: nan,
 61: nan,
 62: nan,
 64: nan,
 66: nan,
 67: nan,
 69: nan,
 7: nan,
 70: nan,
 701: nan,
 703: nan,
 704: nan,
 705: nan,
 72: nan,
 73: nan,
 743: nan,
 744: nan,
 745: nan,
 75: nan,
 751: nan,
 753: nan,
 76: nan,
 77: nan,
 78: nan,
 79: nan,
 8: nan,
 80: 0

In [48]:
mydict_tram = {}
for line in lines:
    mydict_tram[line] = zh[(zh.line_name == line) & (zh.daytype == 0) & (zh.vehicle_type == 'tram')].loc[(zh.time > '17:00:00') & 
                        (zh.time < '18:00:00')].sort_values(by=['occupancy_rate']).occupancy_rate.max()

In [49]:
mydict_tram

{10: 0.5709545454545455,
 11: 0.7039545454545455,
 12: 0.6129545454545454,
 13: 0.7637727272727273,
 14: 0.7272727272727273,
 15: 0.2292727272727273,
 161: nan,
 162: nan,
 163: nan,
 165: nan,
 17: 0.7985909090909091,
 184: nan,
 185: nan,
 2: 0.8863636363636364,
 25: nan,
 29: nan,
 3: 0.7477272727272727,
 301: nan,
 302: nan,
 303: nan,
 304: nan,
 305: nan,
 306: nan,
 307: nan,
 308: nan,
 309: nan,
 31: nan,
 311: nan,
 314: nan,
 315: nan,
 32: nan,
 325: nan,
 33: nan,
 330: nan,
 331: nan,
 341: nan,
 342: nan,
 343: nan,
 344: nan,
 345: nan,
 346: nan,
 347: nan,
 348: nan,
 349: nan,
 35: nan,
 352: nan,
 37: nan,
 38: nan,
 39: nan,
 391: nan,
 392: nan,
 394: nan,
 395: nan,
 396: nan,
 397: nan,
 398: nan,
 4: 0.7955454545454546,
 40: nan,
 411: nan,
 412: nan,
 417: 0.6958636363636364,
 418: 0.7372272727272727,
 419: nan,
 45: nan,
 46: nan,
 5: 0.5204545454545455,
 54: nan,
 6: 0.7038636363636364,
 61: nan,
 62: nan,
 64: nan,
 66: nan,
 67: nan,
 69: nan,
 7: 0.753454

In [50]:
mydict_bus = {}
for line in lines:
    mydict_bus[line] = zh[(zh.line_name == line) & (zh.daytype == 0) & (zh.vehicle_type == 'bus')].loc[(zh.time > '17:00:00') & 
                        (zh.time < '18:00:00')].sort_values(by=['occupancy_rate']).occupancy_rate.max()

In [51]:
mydict_bus

{10: nan,
 11: nan,
 12: nan,
 13: nan,
 14: nan,
 15: nan,
 161: 0.5846250000000001,
 162: 0.284125,
 163: 0.375375,
 165: 0.669625,
 17: nan,
 184: 0.45525,
 185: 0.5705,
 2: nan,
 25: 0.199375,
 29: nan,
 3: nan,
 301: 0.275,
 302: nan,
 303: 0.7375,
 304: nan,
 305: 0.30625,
 306: 0.6775,
 307: 0.193125,
 308: nan,
 309: 0.720875,
 31: nan,
 311: 0.3375,
 314: 0.257875,
 315: 0.1625,
 32: nan,
 325: 0.325,
 33: nan,
 330: nan,
 331: nan,
 341: nan,
 342: nan,
 343: nan,
 344: nan,
 345: nan,
 346: nan,
 347: nan,
 348: nan,
 349: nan,
 35: 0.36200000000000004,
 352: nan,
 37: 0.314375,
 38: 0.177,
 39: 0.08562499999999999,
 391: nan,
 392: nan,
 394: nan,
 395: nan,
 396: nan,
 397: nan,
 398: nan,
 4: nan,
 40: 0.178125,
 411: nan,
 412: nan,
 417: nan,
 418: nan,
 419: nan,
 45: nan,
 46: nan,
 5: nan,
 54: nan,
 6: nan,
 61: 0.813875,
 62: 0.8856249999999999,
 64: 0.288125,
 66: 0.30075,
 67: 0.5625,
 69: 0.504625,
 7: nan,
 70: 0.6599999999999999,
 701: 0.52325,
 703: 0.42975,


In [52]:
mydict_trolleybus = {}
for line in lines:
    mydict_trolleybus[line] = zh[(zh.line_name == line) & (zh.daytype == 0) & (zh.vehicle_type == 'trolleybus')].loc[(zh.time > '17:00:00') & 
                        (zh.time < '18:00:00')].sort_values(by=['occupancy_rate']).occupancy_rate.max()

In [53]:
mydict_trolleybus

{10: nan,
 11: nan,
 12: nan,
 13: nan,
 14: nan,
 15: nan,
 161: nan,
 162: nan,
 163: nan,
 165: nan,
 17: nan,
 184: nan,
 185: nan,
 2: nan,
 25: nan,
 29: nan,
 3: nan,
 301: nan,
 302: nan,
 303: nan,
 304: nan,
 305: nan,
 306: nan,
 307: nan,
 308: nan,
 309: nan,
 31: 0.8204615384615385,
 311: nan,
 314: nan,
 315: nan,
 32: 0.6615384615384615,
 325: nan,
 33: 0.5834615384615384,
 330: nan,
 331: nan,
 341: nan,
 342: nan,
 343: nan,
 344: nan,
 345: nan,
 346: nan,
 347: nan,
 348: nan,
 349: nan,
 35: nan,
 352: nan,
 37: nan,
 38: nan,
 39: nan,
 391: nan,
 392: nan,
 394: nan,
 395: nan,
 396: nan,
 397: nan,
 398: nan,
 4: nan,
 40: nan,
 411: 0.5188461538461538,
 412: 0.2307692307692308,
 417: nan,
 418: nan,
 419: nan,
 45: nan,
 46: 0.4944615384615385,
 5: nan,
 54: nan,
 6: nan,
 61: nan,
 62: nan,
 64: nan,
 66: nan,
 67: nan,
 69: nan,
 7: nan,
 70: nan,
 701: nan,
 703: nan,
 704: nan,
 705: nan,
 72: 0.6090769230769232,
 73: nan,
 743: nan,
 744: nan,
 745: nan,
 

# Results of Analysis

## Mornings

### Morning Overused Lines

Bus: 
- 83
- 89
- 303

Large Bus:
- 80

Tram:
- 6
- S18

In [54]:
# goal is to increase frequency on these lines by 20% in order to be below the benchmark of 85% at all times

Line frequency change:
-	89: Every 15 minutes – 8 in two hours – would need 2 more to get frequency to 12 Minutes
-	83: Every 8 minutes – 15 in two hours – would need 5 more to get to frequency of 6 minutes 
-	80: Every 4 minutes – 30 in two hours – reduce occupancy by 10 percent by increasing 5
-	6: Every 15 minutes – 8 in two hours – need 1 more
-	S18: Every 15 minutes – 2 more
-	303: Every 15 minutes – 2 more


### Morning Underused Lines

Lines and their maximum occupancy

Bus: 
- 162 : 0.362
- 163 : 0.102
- 25 : 0.3
- 305 : 0.26
- 307 : 0.15
- 314 : 0.36
- 38 : 0.1
- 39 : 0.21
- 40 : 0.23
- 705 : 0.37
- 79 : 0.13
- 91 : 0.17
- 918 : 0.21

Tram:
- 15 : 0.29

Line frequency change:
-	163: 10 Minutes – 12 Buses in two hours - Reduce frequency to 13 Minutes and 9 buses
-	38, 91, 918 39, 305: 30 Minutes in two hours – Reduce frequency to 40 minutes and 3 buses each
-	79: 19 Minutes – 6 buses in two hours – Reduce to 30 minutes and 4 buses
-	40: 8 Minutes – 15 buses in two hours – Reduce frequency to 12 minutes and 10 buses
-	705: 15 Minutes – 8 Buses in two hours – Reduce frequency to 20 minutes and 6 buses
- 15 : Every 15 Minutes – 8 in two hours -  Reduce to 24 minutes and 5 trams


## Evenings

### Evening Overused Lines

Bus:
-	62

Tram:
-	2 


Line frequency change:
-	2 : 8 Minutes – 15 trams – increase by 3
-	62 : 8 Minutes – 15 bus – increase by 3

### Evening Underused Lines

Bus: 
- 40 : 0.17 – 8 minutes – reduce from 15 to 10 buses to make 12 minute
- 311: 0.3 – 15 Minutes – reduce from 8 to 6 bus

Large Bus:
- 308 :  0.6

Tram:
- 9 : 0.46 – 8 minutes - reduce from 15 to 10 trams to make 12 minute
- 8 : 0.47


## Measures to take

In order to improve the passenger experience, we are going to reduce the frequency of the underused lines
by 20% in order to increase the frequency of the overused ones by 20%. This will enable us to lower the 
occupancy rate without having to acquire more vehicles. 